### PR TITLE
[Examples] Distributed tensorflow example 

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Adds a self-contained example showing how tensorflow's MultiWorkerMirroredStrategy can be used with SkyPilot to distribute training across multiple nodes.

Our previous example (`resnet_distributed_tf_app.py`) is insufficient because:
1. It uses the python API, which can be harder to grasp new users
2. The training code is from an external repository, and it is hard to follow that code
3. It calls set_inputs and set_outputs, which is unnecessary
4. Does not use workdir or any filemounts

- [x] `sky launch -c myclus tf_distributed.yaml`

### CODE REVIEW

File: examples/tensorflow_distributed/tf_distributed.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    WORKER_LIST="$WORKER_LIST\"$IP:$PORT\","
+  done
+  # Remove trailing comma
+  WORKER_LIST=$(echo $WORKER_LIST | sed 's/,$//')
+  
+  export TF_CONFIG="{\"cluster\": {\"worker\": [$WORKER_LIST]}, \"task\": {\"type\": \"worker\", \"index\": $SKYPILOT_NODE_RANK}}"
+  echo $TF_CONFIG
+  
+  # ======== Run the training script ========
+  python train.py
```
Comment (Michaelvll) [inline examples/tensorflow_distributed/tf_distributed.yaml:53]:
```suggestion
  python train.py
  
```

File: examples/tensorflow_distributed/tf_distributed.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  for IP in $(echo "$SKYPILOT_NODE_IPS" | sed "s/,/ /g")
+  do
+    WORKER_LIST="$WORKER_LIST\"$IP:$PORT\","
+  done
+  # Remove trailing comma
+  WORKER_LIST=$(echo $WORKER_LIST | sed 's/,$//')
+  
+  export TF_CONFIG="{\"cluster\": {\"worker\": [$WORKER_LIST]}, \"task\": {\"type\": \"worker\", \"index\": $SKYPILOT_NODE_RANK}}"
+  echo $TF_CONFIG
+
```
Comment (Michaelvll) [inline examples/tensorflow_distributed/tf_distributed.yaml:51]: Nice! Glad to see this can be done in bash. I am wondering if it is easier to understand for ML people if we do it in python like the following:

```
    python -u <<- EOF
        import json
        import os
        
        port = 2222
        node_ips_str = os.environ.get(SKYPILOT_NODE_IPS)
        node_ips = node_ips_str.strip().split('\n')

        node_rank = os.environ.get(SKYPILOT_NODE_RANK)
        worker_list = [f'{ip}:{port}' for ip in node_ips] 
        tf_config = {
                'cluster': {
                    'worker': worker_list,
                },
                'task': {
                    'type': 'worker',
                    'index': node_rank
                }
            }
        with open('/tmp/{os.environ.get("SKYPILOT_JOB_ID"', 'w') as f:
            json.dump(tf_config, f)
    EOF
    export TF_CONFIG=$(cat /tmp/$SKYPILOT_JOB_ID)
    echo $TF_CONFIG
```

**Review by Michaelvll — APPROVED** [review]
Sorry for missing this PR @romilbhardwaj! I think we should merge this as an example to any potential tensorflow users.

File: examples/tensorflow_distributed/tf_distributed.yaml (hunk 2 above)
Comment (romilbhardwaj) [reply examples/tensorflow_distributed/tf_distributed.yaml]: That's a great idea. Python is definitely easier to understand. I've switched to python in the YAML.

**Comment by romilbhardwaj** [comment]
Thanks for the review @Michaelvll! I've addressed your comments and verified the example works after the updates. Merging now.
